### ISSUE TITLE
diff option should produce a usable patch

### ISSUE THREAD

**Issue by mwichmann** [hakancelik96/unimport#104]
Enhancement request: currently, it makes a pretty display, but if redirected to a file, the "syntax highlighting" makes it unusuable as a patch file: e.g.
```
^[[1;37m--- ExecutorTests.py^[[0m

^[[1;37m+++ ^[[0m

^[[36m@@ -22,8 +22,6 @@^[[0m

 #

 __revision__ = "__FILE__ __REVISION__ __DATE__ __DEVELOPER__"
^[[31m-^[[0m
^[[31m-import sys^[[0m
 import unittest

 import SCons.Executor
```

A common convention is to not emit the highlighting if the output does not appear to be a terminal, another approach might be a flag to generate a patchfile (or files - possibly one-per-modified file) instead of a diff for display.

**Comment by isidentical** [comment]
Absolutely! This is a real issue. Guess we need an extra test if the given diff is appliable or not to verify all future aspects of this feature! I never dealt with the diff feature, but I could say that it shouldn't produce colors at all. There are command-line tools that would highlight any type of the input you are redirecting them (diff is one of that formats), I don't see a value on highlighting diff's by orselves. What do you think about this @hakancelik96?

**Comment by hakancelik96** [comment]
There is such a problem on Windows, the colors do not appear correctly. We will solve this problem.

**Comment by hadialqattan** [comment]
> Absolutely! This is a real issue. Guess we need an extra test if the given diff is appliable or not to verify all future aspects of this feature! I never dealt with the diff feature, but I could say that it shouldn't produce colors at all. There are command-line tools that would highlight any type of the input you are redirecting them (diff is one of that formats), I don't see a value on highlighting diff's by orselves. What do you think about this @hakancelik96?

I think we should use a third party module that can handle the Windows coloring system!

Some suggested modules:
https://github.com/willmcgugan/rich
https://github.com/tartley/colorama

**Comment by hadialqattan** [comment]
> There is such a problem on Windows, the colors do not appear correctly. We will solve this problem.

I think we should include WindowsOS on our CI test suite.

**Comment by hakancelik96** [comment]
@hadialqattan  I agree with you, I tried to add windows OS to the CI test suite but failed, I would be very grateful if anyone added it.

**Comment by hadialqattan** [comment]
> @hadialqattan I agree with you, I tried to add windows OS to the CI test suite but failed, I would be very grateful if anyone added it.

That's because Unimport isn't compatible with the WindowsOS.
Actually, Unimport doesn't support the WindowsOS char encoding system!

I did an experiment, and here are the raw log (cropped):
```log
2020-09-20T14:38:29.0432466Z 
2020-09-20T14:38:29.0432806Z ================================== FAILURES ===================================
2020-09-20T14:38:29.0433204Z ________________________ TestSession.test_bad_encoding ________________________
2020-09-20T14:38:29.0433506Z 
2020-09-20T14:38:29.0433953Z self = <tests.test_session.TestSession testMethod=test_bad_encoding>
2020-09-20T14:38:29.0434381Z 
2020-09-20T14:38:29.0434684Z     def test_bad_encoding(self):
2020-09-20T14:38:29.0435100Z         # Make conflict between BOM and encoding Cookie.
2020-09-20T14:38:29.0435738Z         # https://docs.python.org/3/library/tokenize.html#tokenize.detect_encoding
2020-09-20T14:38:29.0436375Z         bad_encoding = "\ufeff\n# -*- coding: utf-32 -*-\nbad encoding"
2020-09-20T14:38:29.0436840Z >       self.temp_refactor(source=bad_encoding, expected="")
2020-09-20T14:38:29.0437228Z 
2020-09-20T14:38:29.0437529Z tests\test_session.py:33: 
2020-09-20T14:38:29.0437842Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2020-09-20T14:38:29.0438192Z tests\test_session.py:22: in temp_refactor
2020-09-20T14:38:29.0438559Z     tmp.write(source)
2020-09-20T14:38:29.0439024Z C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\tempfile.py:485: in func_wrapper
2020-09-20T14:38:29.0439500Z     return func(*args, **kwargs)
2020-09-20T14:38:29.0440375Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2020-09-20T14:38:29.0440616Z 
2020-09-20T14:38:29.0441217Z self = <encodings.cp1252.IncrementalEncoder object at 0x000002106A83F9B0>
2020-09-20T14:38:29.0441900Z input = '\ufeff\r\n# -*- coding: utf-32 -*-\r\nbad encoding', final = False
2020-09-20T14:38:29.0442216Z 
2020-09-20T14:38:29.0442631Z     def encode(self, input, final=False):
2020-09-20T14:38:29.0443190Z >       return codecs.charmap_encode(input,self.errors,encoding_table)[0]
2020-09-20T14:38:29.0443959Z E       UnicodeEncodeError: 'charmap' codec can't encode character '\ufeff' in position 0: character maps to <undefined>
2020-09-20T14:38:29.0444460Z 
2020-09-20T14:38:29.0444941Z C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\encodings\cp1252.py:19: UnicodeEncodeError
2020-09-20T14:38:29.0445517Z _________________________ TestSession.test_diff_file __________________________
2020-09-20T14:38:29.0445806Z 
2020-09-20T14:38:29.0446278Z self = <tests.test_session.TestSession testMethod=test_diff_file>
2020-09-20T14:38:29.0446694Z 
2020-09-20T14:38:29.0446992Z     def test_diff_file(self):
2020-09-20T14:38:29.0447511Z         with tempfile.NamedTemporaryFile(mode="w", suffix=".py") as tmp:
2020-09-20T14:38:29.0448026Z             tmp.write("import os")
2020-09-20T14:38:29.0448348Z             tmp.seek(0)
2020-09-20T14:38:29.0448764Z             diff_file = self.session.diff_file(path=Path(tmp.name))
2020-09-20T14:38:29.0449354Z             diff = (
2020-09-20T14:38:29.0449664Z                 f"--- {tmp.name}\n",
2020-09-20T14:38:29.0449965Z                 "+++ \n",
2020-09-20T14:38:29.0450227Z                 "@@ -1 +0,0 @@\n",
2020-09-20T14:38:29.0450523Z                 "-import os",
2020-09-20T14:38:29.0450812Z             )
2020-09-20T14:38:29.0451175Z >           self.assertEqual(diff, diff_file)
2020-09-20T14:38:29.0451737Z E           AssertionError: Tuples differ: ('--- C:\\Users\\RUNNER~1\\AppData\\Local\[63 chars] os') != ()
2020-09-20T14:38:29.0452205Z E           
2020-09-20T14:38:29.0452799Z E           First tuple contains 4 additional elements.
2020-09-20T14:38:29.0453207Z E           First extra element 0:
2020-09-20T14:38:29.0453646Z E           '--- C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpggrv3fhb.py\n'
2020-09-20T14:38:29.0454026Z E           
2020-09-20T14:38:29.0454275Z E           + ()
2020-09-20T14:38:29.0454661Z E           - ('--- C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpggrv3fhb.py\n',
2020-09-20T14:38:29.0455032Z E           -  '+++ \n',
2020-09-20T14:38:29.0455294Z E           -  '@@ -1 +0,0 @@\n',
2020-09-20T14:38:29.0455575Z E           -  '-import os')
2020-09-20T14:38:29.0455807Z 
2020-09-20T14:38:29.0456154Z tests\test_session.py:56: AssertionError
2020-09-20T14:38:29.0456595Z ____________________________ TestSession.test_read ____________________________
2020-09-20T14:38:29.0456878Z 
2020-09-20T14:38:29.0457302Z self = <tests.test_session.TestSession testMethod=test_read>
2020-09-20T14:38:29.0457701Z 
2020-09-20T14:38:29.0458001Z     def test_read(self):
2020-09-20T14:38:29.0458309Z         source = "b\ufffdse"
2020-09-20T14:38:29.0458826Z         with tempfile.NamedTemporaryFile(mode="w", suffix=".py") as tmp:
2020-09-20T14:38:29.0459333Z >           tmp.write(source)
2020-09-20T14:38:29.0459567Z 
2020-09-20T14:38:29.0459857Z tests\test_session.py:61: 
2020-09-20T14:38:29.0460168Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2020-09-20T14:38:29.0460614Z C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\tempfile.py:485: in func_wrapper
2020-09-20T14:38:29.0461110Z     return func(*args, **kwargs)
2020-09-20T14:38:29.0461423Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2020-09-20T14:38:29.0461648Z 
2020-09-20T14:38:29.0462134Z self = <encodings.cp1252.IncrementalEncoder object at 0x000002106A858358>
2020-09-20T14:38:29.0462751Z input = 'b\ufffdse', final = False
2020-09-20T14:38:29.0463042Z 
2020-09-20T14:38:29.0463790Z     def encode(self, input, final=False):
2020-09-20T14:38:29.0464664Z >       return codecs.charmap_encode(input,self.errors,encoding_table)[0]
2020-09-20T14:38:29.0465524Z E       UnicodeEncodeError: 'charmap' codec can't encode character '\ufffd' in position 1: character maps to <undefined>
2020-09-20T14:38:29.0466016Z 
2020-09-20T14:38:29.0466497Z C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\encodings\cp1252.py:19: UnicodeEncodeError
2020-09-20T14:38:29.0467043Z =========================== short test summary info ===========================
2020-09-20T14:38:29.0467584Z FAILED tests/test_session.py::TestSession::test_bad_encoding - UnicodeEncodeE...
2020-09-20T14:38:29.0468245Z FAILED tests/test_session.py::TestSession::test_diff_file - AssertionError: T...
2020-09-20T14:38:29.0468904Z FAILED tests/test_session.py::TestSession::test_read - UnicodeEncodeError: 'c...
2020-09-20T14:38:29.0469439Z ================== 3 failed, 114 passed, 6 skipped in 1.86s ===================
2020-09-20T14:38:29.2657381Z ##[error]Process completed with exit code 1.
```
As you see all we have is `UnicodeEncodeError`s!

**Comment by hakancelik96** [comment]
@hadialqattan You can solve Windows problems then you can add windows configuring to [test.yaml](https://github.com/hakancelik96/unimport/blob/master/.github/workflows/test.yml). file and send a PR.

And finally, we support windows in the new version.

**Comment by hakancelik96** [comment]
Do you want to solve this issue, I can assign to you @hadialqattan

**Comment by hadialqattan** [comment]
> Do you want to solve this issue, I can assign to you @hadialqattan

This issue will be closed when I solve this issue ~> #110

**Comment by hakancelik96** [comment]
Instead of using 3rd party lib to run the color feature in windows, we can do it like [colorama](https://github.com/tartley/colorama) and [Pre-commit](https://github.com/pre-commit/pre-commit), I leave the addresses for your review.

https://github.com/pre-commit/pre-commit/blob/master/pre_commit/color.py

https://github.com/tartley/colorama/blob/master/colorama/win32.py
